includes/class-wc-stripe-intent-controller.php

```diff
@@ -676,6 +676,152 @@ public function maybe_process_upe_redirect() {
 			$gateway->maybe_process_upe_redirect();
 		}
 	}
-}
 
-new WC_Stripe_Intent_Controller();
+	/**
+	 * Creates and confirm a payment intent with the given payment information.
+	 * Used for dPE.
+	 *
+	 * @param object $payment_information The payment information needed for creating and confirming the intent.
+	 *
+	 * @throws WC_Stripe_Exception - If the create intent call returns with an error.
+	 *
+	 * @return array
+	 */
+	public function create_and_confirm_payment_intent( $payment_information ) {
+		// Throws a WC_Stripe_Exception if required information is missing.
+		$this->validate_create_and_confirm_intent_payment_information( $payment_information );
+
+		$order                 = $payment_information['order'];
+		$selected_payment_type = $payment_information['selected_payment_type'];
+		$payment_method_types  = $this->get_payment_method_types_for_intent_creation( $selected_payment_type, $order->get_id() );
+
+		$request = [
+			'amount'               => $payment_information['amount'],
+			'capture_method'       => $payment_information['capture_method'],
+			'confirm'              => 'true',
+			'currency'             => $payment_information['currency'],
+			'customer'             => $payment_information['customer'],
+			/* translators: 1) blog name 2) order number */
+			'description'          => sprintf( __( '%1$s - Order %2$s', 'woocommerce-gateway-stripe' ), wp_specialchars_decode( get_bloginfo( 'name' ), ENT_QUOTES ), $order->get_order_number() ),
+			'metadata'             => $payment_information['metadata'],
+			'payment_method'       => $payment_information['payment_method'],
+			'payment_method_types' => $payment_method_types,
+			'statement_descriptor' => $payment_information['statement_descriptor'],
+		];
+
+		if ( $payment_information['save_payment_method_to_store'] ) {
+			$request['setup_future_usage'] = 'off_session';
+		}
+
+		$payment_intent = WC_Stripe_API::request_with_level3_data(
+			$request,
+			'payment_intents',
+			$payment_information['level3'],
+			$order
+		);
+
+		// Only update the payment_type if we have a reference to the payment type the customer selected.
+		if ( '' !== $selected_payment_type ) {
+			$order->update_meta_data( '_stripe_upe_payment_type', $selected_payment_type );
+		}
+
+		$order->update_status( 'pending', __( 'Awaiting payment.', 'woocommerce-gateway-stripe' ) );
```

Comment:
I haven't tested anything yet but I'm curious why we need to update the status of the order to pending when dealing with deferred intents. What is the order status before this?

I'm wondering if we should have a `$order->needs_payment()` somewhere so that we're only attempting to create and confirm an intent for orders that have pending/failed status.

Maybe inside `process_payment_with_deferred_intent()`?

Reply:
> I haven't tested anything yet but I'm curious why we need to update the status of the order to pending when dealing with deferred intents. What is the order status before this?

You're right, no need to update the order status. The order is already "pending" at this point. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2753/commits/fef7018fddf43487b8205e46ed3ad4145e8ae289

> I'm wondering if we should have a $order->needs_payment() somewhere so that we're only attempting to create and confirm an intent for orders that have pending/failed status.

Yeah, I agree. 

I see there's a more complex logic around this in [verify_intent_after_checkout()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/class-wc-gateway-stripe.php#L823), but I think it's specific for the legacy PE implementation because it runs between the Checkout and Thank you pages.

Would it be okay with you to have an issue for addressing this, instead of doing it in this PR? I'm thinking that webhooks and cache may also come into play, but would like to dive into it. Also to not delay merging this base PR.

Reply:
> Would it be okay with you to have an issue for addressing this, instead of doing it in this PR? I'm thinking that webhooks and cache may also come into play, but would like to dive into it. Also to not delay merging this base PR.

That sounds good to me?

Reply:
Yay! GH issue created over here https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2785 . Of course feel free to update it!

And thanks as always for pointing these out 😄